In [1]:
%pip install -Uq langchain langchain-core langchain-groq

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 25.0.1 -> 26.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [4]:
from langchain.agents import create_agent
from langchain_core.tools import tool
from dotenv import load_dotenv
from langchain_groq import ChatGroq

@tool
def get_weather(city: str) ->str:
    '''도시 이름을 입력받아 현재 날씨 정보를 반환합니다
    
    Args:
        city: 도시 이름
        
    Returns:
        str: 현재 도시 날씨정보를 반환합니다.
    '''

    return f'{city}의 현재 날씨는 맑고 25도입니다'

openai_llm_model = ChatGroq(
    model="openai/gpt-oss-20b",
    temperature=0.7
)

agent_exp = create_agent(
    model=openai_llm_model,
    tools=[get_weather]
)

response = agent_exp.invoke({
    "messages":[
        {
            "role":"user",
            "content":"서울 날씨 알려줘"
        }
    ]
})

print(response)

{'messages': [HumanMessage(content='서울 날씨 알려줘', additional_kwargs={}, response_metadata={}, id='63a3b7c6-6e25-4e24-8939-88212396a526'), AIMessage(content='', additional_kwargs={'reasoning_content': 'User: "서울 날씨 알려줘" (Tell me the weather in Seoul). Need to call get_weather with city "Seoul".', 'tool_calls': [{'id': 'fc_5e2972d6-9666-4a85-95d1-a93a484cd504', 'function': {'arguments': '{"city":"Seoul"}', 'name': 'get_weather'}, 'type': 'function'}]}, response_metadata={'token_usage': {'completion_tokens': 53, 'prompt_tokens': 159, 'total_tokens': 212, 'completion_time': 0.068575582, 'completion_tokens_details': {'reasoning_tokens': 29}, 'prompt_time': 0.010015648, 'prompt_tokens_details': None, 'queue_time': 0.083025666, 'total_time': 0.07859123}, 'model_name': 'openai/gpt-oss-20b', 'system_fingerprint': 'fp_1074f9ce08', 'service_tier': 'on_demand', 'finish_reason': 'tool_calls', 'logprobs': None, 'model_provider': 'groq'}, id='lc_run--019d2e90-c462-7240-8bff-eb9a8b1cd51b-0', tool_calls=

{'messages': [HumanMessage(content='서울 날씨 알려줘', additional_kwargs={}, response_metadata={}, id='63a3b7c6-6e25-4e24-8939-88212396a526'), 
AIMessage(content='', additional_kwargs={'reasoning_content': 'User: "서울 날씨 알려줘" (Tell me the weather in Seoul). Need to call get_weather with city "Seoul".', 'tool_calls': [{'id': 'fc_5e2972d6-9666-4a85-95d1-a93a484cd504', 'function': {'arguments': '{"city":"Seoul"}', 'name': 'get_weather'}, 'type': 'function'}]}, response_metadata={'token_usage': {'completion_tokens': 53, 'prompt_tokens': 159, 'total_tokens': 212, 'completion_time': 0.068575582, 'completion_tokens_details': {'reasoning_tokens': 29}, 'prompt_time': 0.010015648, 'prompt_tokens_details': None, 'queue_time': 0.083025666, 'total_time': 0.07859123}, 'model_name': 'openai/gpt-oss-20b', 'system_fingerprint': 'fp_1074f9ce08', 'service_tier': 'on_demand', 'finish_reason': 'tool_calls', 'logprobs': None, 'model_provider': 'groq'}, id='lc_run--019d2e90-c462-7240-8bff-eb9a8b1cd51b-0', tool_calls=[{'name': 'get_weather', 'args': {'city': 'Seoul'}, 'id': 'fc_5e2972d6-9666-4a85-95d1-a93a484cd504', 'type': 'tool_call'}], invalid_tool_calls=[], usage_metadata={'input_tokens': 159, 'output_tokens': 53, 'total_tokens': 212, 'output_token_details': {'reasoning': 29}}), 

ToolMessage(content='Seoul의 현재 날씨는 맑고 25도입니다', name='get_weather', id='bd2bdcd1-748a-4bd3-b39a-07d881d1d590', tool_call_id='fc_5e2972d6-9666-4a85-95d1-a93a484cd504'), 

AIMessage(content='서울의 현재 날씨는 맑고 25도입니다.', additional_kwargs={'reasoning_content': 'User asked: "서울 날씨 알려줘" We responded with the weather. Now we should respond in Korean with the weather info. Probably just give the weather.'}, response_metadata={'token_usage': {'completion_tokens': 57, 'prompt_tokens': 199, 'total_tokens': 256, 'completion_time': 0.074317938, 'completion_tokens_details': {'reasoning_tokens': 34}, 'prompt_time': 0.010003599, 'prompt_tokens_details': None, 'queue_time': 0.019519198, 'total_time': 0.084321537}, 'model_name': 'openai/gpt-oss-20b', 'system_fingerprint': 'fp_24bfb4a850', 'service_tier': 'on_demand', 'finish_reason': 'stop', 'logprobs': None, 'model_provider': 'groq'}, id='lc_run--019d2e90-c5ee-78a2-9342-8e155008b1de-0', tool_calls=[], invalid_tool_calls=[], usage_metadata={'input_tokens': 199, 'output_tokens': 57, 'total_tokens': 256, 'output_token_details': {'reasoning': 34}})]}

In [9]:
print(response['messages'][-1].content)

서울의 현재 날씨는 맑고 25도입니다.
